### PR TITLE
add RtcState object

### PR DESCRIPTION
**Pull request by lauralt**
This PR is adding support for saving and restoring the state of the RTC device. We are starting to add this support in rust-vmm to enable snapshot use cases, such as live migration, and to unblock the consumption of components (e.g. virtio devices, legacy devices).

This is also converting vm-superio to a workspace, in which the base `vm-superio` provides the functionality required to get/set the state of the device, and `vm-superio-ser` defines the `RtcStateSer` object (and in the future the `...StateSer` objects for the other devices). `RtcStateSer` mirrors the state structure from `vm-superio`, and derives the required (de)serialization/versioning traits (i.e. `Serialize`, `Deserialize`,` Versionize`). If we later change the state of the component, this one will have to be updated as well (by also adding the required version constraints on it).

### CODE REVIEW

**Review by alexandruag — APPROVED** [review]

File: src/rtc_pl031.rs [hunk 1]
```diff
@@ -147,52 +161,77 @@ fn get_current_time() -> u32 {
...
+    /// Creates a new `AMBA PL031 RTC` instance without any metric
+    /// capabilities.
+    pub fn new() -> Self {
+        Self::from_state(&RtcState::default(), NoEvents)
+    }
+}
+
+impl<EV: RtcEvents> Rtc<EV> {
+    /// Creates a new `AMBA PL031 RTC` instance from a given `state` and invokes the `rtc_events`
+    /// implementation of `RtcEvents` during operation.
```
Comment (andreeaflorescu) [inline src/rtc_pl031.rs:198]: I think we should also add a comment about wanting to start from a clean state, in which case people should use `with_events`.

File: src/rtc_pl031.rs [hunk 2]
```diff
@@ -147,52 +161,77 @@ fn get_current_time() -> u32 {
...
+            ris: state.ris,
             // A struct implementing RtcEvents for tracking the occurrence of
             // significant events.
-            events: rtc_events,
+            events,
+        }
+    }
+
+    /// Creates a new `AMBA PL031 RTC` instance and invokes the `rtc_events`
+    /// implementation of `RtcEvents` during operation.
```
Comment (andreeaflorescu) [inline src/rtc_pl031.rs:218]: The part about "invokes the `rtc_events`" sounds like a implementation detail, and it shouldn't be part of the interface documentation. How about: "Creates a new `AMBA PL031 RTC` instance that is able to track events during operation using the passed `rtc_events` object." We should also add a paragraph about the state being the default one or something like that.

File: src/rtc_pl031.rs [hunk 3]
```diff
@@ -854,4 +893,53 @@ mod tests {
...
+        let second_read = u32::from_le_bytes(data);
+        // The RTC values should be different.
+        assert!(second_read > first_read);
+
+        // Reading from the LR register should return the same value as before saving the state.
+        rtc_from_state.read(RTCLR, &mut data2);
+        assert_eq!(data, data2);
+
+        // Check that the restored `Rtc` doesn't keep the state of the old `metrics` object.
+        assert_eq!(rtc_from_state.events.invalid_write_count.count(), 0);
```
Comment (andreeaflorescu) [inline src/rtc_pl031.rs:943]: nit: For the sake of the example, maybe it would be a good idea to have a "test as an example" where the metrics are saved as well.

File: crates/vm-superio/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "vm-superio"
+version = "0.3.0"
+description = "Emulation for legacy devices"
+keywords = ["superio", "emulation"]
+repository = "https://github.com/rust-vmm/vm-superio"
+readme = "README.md"
```
Comment (andreeaflorescu) [inline crates/vm-superio/Cargo.toml:7]: This is not going to be valid anymore when we publish the crate on crates.io because it will not find the README.md file in the root of the crate. The root of the crate is where `Cargo.toml` is defined as I understand it. We should see how we need to address this.

File: crates/vm-superio/src/rtc_pl031.rs [hunk 5]
```diff
@@ -210,7 +210,7 @@ impl<EV: RtcEvents> Rtc<EV> {
             ris: state.ris,
             // A struct implementing RtcEvents for tracking the occurrence of
             // significant events.
-            events,
+            events
```
Comment (andreeaflorescu) [inline crates/vm-superio/src/rtc_pl031.rs]: nit: this should not be part of the commit.

File: crates/vm-superio-ser/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,16 @@
+[package]
+name = "vm-superio-ser"
+version = "0.1.0"
```
Comment (andreeaflorescu) [inline crates/vm-superio-ser/Cargo.toml:3]: I think we sort of missed talking about this during the design phase, but how do we think about versions? We should have a way to specify what version of vm-superio-ser is compatible with what version of vm-superio.

File: crates/vm-superio-ser/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+readme = "README.md"
+authors = ["rust-vmm AWS maintainers <rust-vmm-maintainers@amazon.com>"]
+license = "Apache-2.0 OR BSD-3-Clause"
+edition = "2018"
+
+[dependencies]
+serde = { version = ">=1.0.27", features = ["derive"] }
+versionize = ">=0.1.6"
+versionize_derive = ">=0.1.3"
+vm-superio = { path =  "../vm-superio" }
```
Comment (andreeaflorescu) [inline crates/vm-superio-ser/Cargo.toml:16]: We will not be able to publish it with path dependencies. We need to specify the dependency similar to how serde does it: https://github.com/serde-rs/serde/blob/a6690ea2fe83924e5bb37bbb6a1341444d26a65b/serde/Cargo.toml#L17

File: src/rtc_pl031.rs (hunk 3 above)
Comment (lauralt) [reply src/rtc_pl031.rs:943]: Added an example, it's a bit weird though :-?.

File: crates/vm-superio-ser/Cargo.toml (hunk 6 above)
Comment (lauralt) [reply crates/vm-superio-ser/Cargo.toml:3]: Added a new section in the README, hopefully it makes sense.

File: crates/vm-superio-ser/Cargo.toml (hunk 7 above)
Comment (lauralt) [reply crates/vm-superio-ser/Cargo.toml:16]: Thanks, I think we'll be able to update it once we have a release of vm-superio with the RtcState.

File: crates/vm-superio/src/rtc_pl031.rs (hunk 5 above)
Comment (lauralt) [reply crates/vm-superio/src/rtc_pl031.rs]: removed it.

File: crates/vm-superio/Cargo.toml (hunk 4 above)
Comment (lauralt) [reply crates/vm-superio/Cargo.toml:7]: I added symlinks similar to how `serde` does it. From what I understand, it won't work with relative paths. [This release](https://crates.io/crates/serde/0.9.15) used relative paths, and it doesn't look right (https://github.com/serde-rs/serde/blob/v0.9.15/serde/Cargo.toml#L12). v1.0.0 used symlinks: https://github.com/serde-rs/serde/tree/v1.0.0/serde, https://crates.io/crates/serde/1.0.0.

File: crates/vm-superio/src/rtc_pl031.rs [hunk 8]
```diff
@@ -134,6 +134,20 @@ pub struct Rtc<EV: RtcEvents> {
     events: EV,
 }
 
+/// The state of the Rtc device.
+pub struct RtcState {
```
[the remark this answers is not in the document]
Comment (andreeaflorescu) [reply crates/vm-superio/src/rtc_pl031.rs:138]: I think it might also help to have a derive Default, but I am not sure what should be the implementation of that. Maybe we can implement default ourselves and have everything as 0s expect the LR which we can initialize as the current time on host.

[the remark this answers is not in the document]
Comment (lauralt) [reply crates/vm-superio/src/rtc_pl031.rs:138]: Default is already implemented for RtcState. I was using `derive` initially, but switched to a manual implementation to be on the safe side, and it also helped to have a clearer documentation.

[the remark this answers is not in the document]
Comment (lauralt) [reply crates/vm-superio/src/rtc_pl031.rs:138]: Added derives for Debug, Clone and PartialEq. I also realized now that `RtcStateSer` was deriving Default. I removed the derive, we can add a manual implementation if we think it's useful, but it doesn't seem so, at least for now.

File: crates/vm-superio/src/rtc_pl031.rs [hunk 9]
```diff
@@ -147,55 +161,82 @@ fn get_current_time() -> u32 {
...
             // The raw interrupt is initialised as not asserted.
             ris: 0,
+        }
+    }
+}
+
+impl Rtc<NoEvents> {
+    /// Creates a new `AMBA PL031 RTC` instance without any metric capabilities. The instance is
+    /// created from the default state.
+    pub fn new() -> Self {
```
[the remark this answers is not in the document]
Comment (lauralt) [reply crates/vm-superio/src/rtc_pl031.rs:191]: As discussed offline, we can postpone this a bit until we have a nice idea to avoid weird naming.

File: src/rtc_pl031.rs [hunk 10]
```diff
@@ -854,4 +896,63 @@ mod tests {
...
+
+        // Let's increment again a metric, and this time save the state of events as well (separate
+        // from the base state).
+        // Do an invalid `write` in order to increment a metric.
+        let data3 = 123u32.to_le_bytes();
+        rtc_from_state.write(AMBA_ID_HIGH + 4, &data3);
+        assert_eq!(rtc_from_state.events.invalid_write_count.count(), 1);
+
+        let state2 = rtc_from_state.state();
+        let metrics2 = metrics;
```
Comment (andreeaflorescu) [inline src/rtc_pl031.rs:952]: Maybe we can rename this to `saved_metrics` and add a comment here that we are mimicking saving the metrics for the sake of the example so that it is a bit less weird :laughing:

File: crates/vm-superio-ser/src/rtc_pl031.rs [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+            lr: state.lr,
+            offset: state.offset,
+            mr: state.mr,
+            imsc: state.imsc,
+            ris: state.ris,
+        }
+    }
+}
+
+impl RtcStateSer {
```
Comment (andreeaflorescu) [inline crates/vm-superio-ser/src/rtc_pl031.rs:43]: Is there a reason for implementing `new` here, and from for the reversed operation?

Comment (lauralt) [reply crates/vm-superio-ser/src/rtc_pl031.rs:43]: Hmm, not really, I think I was under the impression I can't implement `From` here. I replaced with `From`.

File: crates/vm-superio/src/rtc_pl031.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (lauralt) [reply crates/vm-superio/src/rtc_pl031.rs:139]: I also added a `Default` impl, it would look a bit awkward to convert with `From` from an RtcState::default() in the consumer code in case a default RtcStateSer is needed.

File: src/rtc_pl031.rs (hunk 10 above)
Comment (lauralt) [reply src/rtc_pl031.rs:952]: done.

**Review by andreeaflorescu — APPROVED** [review]

**Review by alexandruag — APPROVED** [review]
